edb/schema/inheriting.py

```diff
@@ -389,13 +495,22 @@ class AlterInherit(sd.Command):
     astnode = qlast.AlterAddInherit, qlast.AlterDropInherit
 
     @classmethod
-    def _cmd_tree_from_ast(cls, schema, astnode, context):
+    def _cmd_tree_from_ast(
+        cls,
+        schema: s_schema.Schema,
+        astnode: Any,
```

Comment:
`qlast.ObjectDDL`

Reply:
@elprans, can we use a generic like in the solution I propose in my last commit? I tried this way because using exactly ObjectDDL was then causing problems in the annotations in `annos` module, and I understand that subclasses will always be called with their specific ast node (?).

By the way, here and in the other places where I left "Any", I did exactly because I needed your input (not because I was thinking it was right to leave Any).

Reply:
> causing problems in the annotations in annos module, and I understand that subclasses will always be called with their specific ast node (?).

Yes, and the subclasses should be called with a subclass of `ObjectDDL`, so there should be no problem.  A generic is overkill here, since you're only annotating a single argument.  Can you show the actual error you're seeing?

Reply:
`Argument 2 of "_cmd_tree_from_ast" is incompatible with supertype "CreateInheritingObject"; supertype defines the argument type as "ObjectDDL"`

![image](https://user-images.githubusercontent.com/2576032/72824547-c01a2300-3c75-11ea-82c2-8f8600f1bd72.png)

I know it could be resolved using qlast.ObjectDDL and assert in the function body that it's an instance of the expected type, but this way looks also not optimal to me.

Reply:
@ambv Aren't overloaded methods in subtypes supposed to be covariant with respect to the arguments?

Reply:
Maybe there is a better way in mypy.

FWIW, in C# this would be a normal case for a generic, but I agree that in Python generics looks more like "heavy machinery".

```csharp
  public class ObjectDDL
  {

  }

  public class SpecificObjectDDL : ObjectDDL
  {
    public string Foo { get; set; }
  }

  public class Example<T> where T : ObjectDDL
  {
    public virtual void Something(T arg)
    {

    }
  }

  public class SpecificExample : Example<SpecificObjectDDL>
  {
    public override void Something(SpecificObjectDDL arg)
    {
      // throw exception if arg is null (not mandatory)
      Console.WriteLine(arg.Foo);
    }
  }
```

Reply:
Yes, I always forget, this is the Liskov Substitution Principle.  The explicit use of a covariant generic is needed in the parent class.

Reply:
In conclusion, what pattern do you prefer in this case? Do you prefer subclasses to have the generic annotation (ObjectDDL) and have assert isinstance?

Reply:
The solution in your last commit is correct, Roberto.